### PR TITLE
GH-33960: [C++] Add DeclarationToSchema and DeclarationToString helper methods.

### LINKED ISSUE #33960: [R] Output schema for aggregation is sometimes innacurate
### Describe the bug, including details regarding any error messages, version, and platform.

Discovered in https://github.com/duckdb/duckdb/issues/5895 and ended up in 32-bit integers (our calculated output schema) interpreted as 64-bit integers (the actual resulting schema).

In 11.0.0 you can now get a RecordBatchReader from an Arrow dplyr query without starting to evaluate the query, so `as_record_batch_reader(query)$schema` is safe. We might be able to replace our current schema inference with that except we'll need to be careful about performance because that process involves building up the entire query from scratch including any nested components. I don't know how often we request that schema but maybe we can request it less as part of this PR.

An example where our calculated output is incorrect:

``` r
library(arrow, warn.conflicts = FALSE)
#> Some features are not enabled in this build of Arrow. Run `arrow_info()` for more information.
library(dplyr, warn.conflicts = FALSE)

small <- tibble(a = c(letters, letters)) |> as_arrow_table()

query <- small |> count(a)
query  
#> Table (query)
#> a: string
#> n: int32
#> 
#> See $.data for the source Arrow object

collect(query, as_data_frame = FALSE)
#> Table
#> 26 rows x 2 columns
#> $a <string>
#> $n <int64>
as_record_batch_reader(query)$schema
#> Schema
#> a: string
#> n: int64
```

<sup>Created on 2023-01-31 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### Component(s)

R

### PR DESCRIPTION
**Pull request by westonpace**
Also cleans up the Acero server example to use the DeclarationToXyz methods
* Closes: #33960

### CODE REVIEW

**Comment by westonpace** [comment]
I'm not sure if this fully closes #33960 but it was inspired by it.

File: cpp/src/arrow/compute/exec/exec_plan.cc [hunk 1]
```diff
@@ -822,6 +868,10 @@ Result<std::unique_ptr<RecordBatchReader>> DeclarationToReader(
     }
 
     Status Close() override {
+      if (!!iterator_) {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/exec_plan.cc:871]: Should this just be `!iterator_`?

Comment (westonpace) [reply cpp/src/arrow/compute/exec/exec_plan.cc]: :cold_sweat: yes.  Fixed.

**Review by lidavidm — APPROVED** [review]

**Comment by ianmcook** [comment]
#33917 provides an immediate fix for #33960. I unlinked this from #33917 since it does not.

**Comment by westonpace** [comment]
> https://github.com/apache/arrow/pull/33917 provides an immediate fix for https://github.com/apache/arrow/issues/33960. I unlinked this from https://github.com/apache/arrow/issues/33960 since it does not. Nope, the bots wouldn't let me do that without changing the PR title, I think.

I thought the merge script would allow me to merge this without updating the issue but it seems to have closed it anyway (I suppose GH does that automatically).  I reopened #33960 manually.

**Comment by lidavidm** [comment]
In this case, the PR title and description still reference that issue and the sidebar notes that the issue is linked, so yeah, it seems GitHub did that

**Pull request merged**